modules/llrt_crypto/src/subtle/generate_key.rs

```diff
@@ -0,0 +1,276 @@
+use rsa::{
+    pkcs1::EncodeRsaPrivateKey,
+    {rand_core::OsRng, BigUint, RsaPrivateKey},
+};
+
+use crate::{
+    subtle::{extract_sha_hash, CryptoNamedCurve, KeyGenAlgorithm, Sha},
+    SYSTEM_RANDOM,
+};
+
+use super::crypto_key::{CryptoKey, CryptoKeyPair};
+
+static PUB_EXPONENT_1: OnceLock<BigUint> = OnceLock::new();
+static PUB_EXPONENT_2: OnceLock<BigUint> = OnceLock::new();
+
+static SYMMETRIC_USAGES: &[&str] = &["encrypt", "decrypt", "wrapKey", "unwrapKey"];
+static SIGNATURE_USAGES: &[&str] = &["sign", "verify"];
+static EMPTY_USAGES: &[&str] = &[];
+static SIGN_USAGES: &[&str] = &["sign"];
+static RSA_OAEP_USAGES: &[&str] = &["decrypt", "unwrapKey"];
+static ECDH_USAGES: &[&str] = &["deriveKey", "deriveBits"];
+
+static SUPPORTED_USAGES_ARRAY: &[(&str, &[&str])] = &[
+    ("AES-CBC", SYMMETRIC_USAGES),
+    ("AES-CTR", SYMMETRIC_USAGES),
+    ("AES-GCM", SYMMETRIC_USAGES),
+    ("ECDH", ECDH_USAGES),
+    ("ECDSA", SIGNATURE_USAGES),
+    ("HMAC", SIGNATURE_USAGES),
+    ("RSA-OAEP", SYMMETRIC_USAGES),
+    ("RSA-PSS", SIGNATURE_USAGES),
+    ("RSASSA-PKCS1-v1_5", SIGNATURE_USAGES),
+];
+
+static MANDATORY_USAGES_ARRAY: &[(&str, &[&str])] = &[
+    ("AES-CBC", EMPTY_USAGES),
+    ("AES-CTR", EMPTY_USAGES),
+    ("AES-GCM", EMPTY_USAGES),
+    ("ECDH", ECDH_USAGES),
+    ("ECDSA", SIGN_USAGES),
+    ("HMAC", EMPTY_USAGES),
+    ("RSA-OAEP", RSA_OAEP_USAGES),
+    ("RSA-PSS", SIGN_USAGES),
+    ("RSASSA-PKCS1-v1_5", SIGN_USAGES),
+];
+
+pub async fn subtle_generate_key<'js>(
+    ctx: Ctx<'js>,
+    algorithm: Value<'js>,
+    extractable: bool,
+    key_usages: Array<'js>,
+) -> Result<Value<'js>> {
+    let (name, key_gen_algorithm) = extract_generate_key_algorithm(&ctx, &algorithm)?;
+
+    let (private_usages, public_usages) = classify_and_check_usages(&ctx, &name, &key_usages)?;
+
+    let bytes = generate_key(&ctx, &key_gen_algorithm)?;
+
+    if name.starts_with("AES") || name == "HMAC" {
+        CryptoKey::new(
+            ctx.clone(),
+            "secret".to_string(),
+            extractable,
+            algorithm,
+            public_usages.into_js(&ctx)?.into_array().unwrap(),
+            &bytes,
+        )
+        .into_js(&ctx)
+    } else {
+        let private_key = CryptoKey::new(
+            ctx.clone(),
+            "private".to_string(),
+            extractable,
+            algorithm.clone(),
+            private_usages.into_js(&ctx)?.into_array().unwrap(),
+            &bytes,
+        )?;
+        let private_key = Class::instance(ctx.clone(), private_key)?;
+        let public_key = CryptoKey::new(
+            ctx.clone(),
+            "public".to_string(),
+            true,
+            algorithm,
+            public_usages.into_js(&ctx)?.into_array().unwrap(),
+            &bytes,
+        )?;
+        let public_key = Class::instance(ctx.clone(), public_key)?;
+        CryptoKeyPair::new(ctx.clone(), private_key, public_key).into_js(&ctx)
```

Comment:
Looking how chrome does it, we can actually remove this class and just return an object with two properties. This should not be a class. I know it was an object before but I missed this completely.
Not that it should have readonly properties and not writable properties.
```rust
let mut key_pair = Object::new(ctx)?;
key_pair.prop("privateKey", public_key)?;
key_pair.prop("publicKey", private_key)?;
Ok(key_pair);
```

Reply:
There doesn't seem to be any problem functionally, but the display of console.log is slightly inconvenient. When logging CryptoKeyPair in bun, the internal contents are displayed, but in LLRT, they are not displayed.

Is it because the specifications of the console module are different?

```javascript
const key = await crypto.subtle
  .generateKey(
    {
      name: "ECDSA",
      namedCurve: "P-256",
    },
    true,
    ["sign", "verify"]
  );

console.log("console.log(key):");
console.log(key);
console.log("console.log(key.publicKey):");
console.log(key.publicKey);
```

```
% llrt subtleGenerateKey.js
console.log(key):
{}
console.log(key.publicKey):
{
  algorithm: {
    name: 'ECDSA',
    namedCurve: 'P-256'
  },
  type: 'public',
  usages: [ 'verify' ],
  extractable: true
}

% bun subtleGenerateKey.js 
console.log(key):
{
  privateKey: CryptoKey {
    type: "private",
    extractable: true,
    algorithm: {
      name: "ECDSA",
      namedCurve: "P-256",
    },
    usages: [ "sign" ],
  },
  publicKey: CryptoKey {
    type: "public",
    extractable: true,
    algorithm: {
      name: "ECDSA",
      namedCurve: "P-256",
    },
    usages: [ "verify" ],
  },
}
console.log(key.publicKey):
CryptoKey {
  type: "public",
  extractable: true,
  algorithm: {
    name: "ECDSA",
    namedCurve: "P-256",
  },
  usages: [ "verify" ],
}
```

Reply:
> I would keep the class. There is a readonly flag on the derive of class. It basically is the same as a js object but we can extend it.

@Sytten , Are you saying that it can fulfill both what you're saying and what @richarddavison  is saying?

Reply:
For console log, you have to make the prop enumerable. Search for this and see example of how to in the project 👍